pkg/operator/ceph/object/zone/controller.go

```diff
@@ -0,0 +1,236 @@
+		if !cephObjectStoreZone.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Remove finalizer
+			err = opcontroller.RemoveFinalizer(r.client, cephObjectStoreZone)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+			}
+
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		return reconcileResponse, nil
+	}
+
+	// Set a finalizer so we can do cleanup before the object goes away
+	err = opcontroller.AddFinalizerIfNotPresent(r.client, cephObjectStoreZone)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
+	}
+
+	// DELETE: the CR was deleted
+	if !cephObjectStoreZone.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("deleting zone CR %q", cephObjectStoreZone.Name)
+
+		// Remove finalizer
+		err = opcontroller.RemoveFinalizer(r.client, cephObjectStoreZone)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// validate the zone settings
+	err = ValidateZoneCR(cephObjectStoreZone)
+	if err != nil {
+		cephObjectStoreZone.Status.Phase = k8sutil.ReconcileFailedStatus
+		err := opcontroller.UpdateStatus(r.client, cephObjectStoreZone)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+		return reconcile.Result{}, errors.Wrapf(err, "invalid name CR %q spec", cephObjectStoreZone.Name)
+	}
+
+	// Start object reconciliation, updating status for this
+	cephObjectStoreZone.Status.Phase = k8sutil.ReconcilingStatus
+	err = opcontroller.UpdateStatus(r.client, cephObjectStoreZone)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+	}
+
+	// CREATE/UPDATE CEPH ZONE
+	reconcileResponse, err = r.reconcileCephZone(cephObjectStoreZone)
+	if err != nil {
+		cephObjectStoreZone.Status.Phase = k8sutil.ReconcileFailedStatus
+		errStatus := opcontroller.UpdateStatus(r.client, cephObjectStoreZone)
+		if errStatus != nil {
+			logger.Errorf("failed to set status. %v", errStatus)
+		}
+		return reconcileResponse, err
+	}
+
+	// Set Ready status, we are done reconciling
+	cephObjectStoreZone.Status.Phase = k8sutil.ReadyStatus
+	err = opcontroller.UpdateStatus(r.client, cephObjectStoreZone)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+	}
+
+	// Return and do not requeue
+	logger.Debug("done reconciling")
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileObjectStoreZone) reconcileCephZone(cephObjectStoreZone *cephv1.CephObjectStoreZone) (reconcile.Result, error) {
+	err := createCephZone(r.context, cephObjectStoreZone.Name, cephObjectStoreZone.Namespace, cephObjectStoreZone.Spec.Realm, cephObjectStoreZone.Spec.ZoneGroup, cephObjectStoreZone.Spec.IsMaster)
```

Comment:
Why isn't `createCephZone` a method of `ReconcileObjectStoreZone`? If so you wouldn't need to pass the context.